### PR TITLE
[refactor] #2253: Add `Registrable` trait to `data_model`

### LINKED ISSUE #2253: `data_model` builder trait
In our `data_model` we have a lot of builders (though only for the `Identifiable` structures) offering the `build` method as was noted [here](https://github.com/hyperledger/iroha/pull/2211#discussion_r877841198). Currently, we have an `Identifiable::RegisteredWith` which connects the builder and the structure to build. Most likely we can have both

### PR DESCRIPTION
**Pull request by Erigara**
Signed-off-by: Shanin Roman <shanin1000@yandex.ru>

### Description of the Change

Introduce `Registrable` trait to create unified interface for building objects.

### Issue

Closes #2253.

### Benefits

Unified interface.

### Possible Drawbacks

None.

### CODE REVIEW

File: data_model/src/asset.rs [hunk 1]
```diff
@@ -416,8 +416,24 @@ pub struct NewAssetDefinition {
     metadata: Metadata,
 }
 
+impl Builder for NewAssetDefinition {
+    type Target = AssetDefinition;
+
+    #[cfg(feature = "mutable_api")]
```
Comment (mversic) [inline data_model/src/asset.rs:422]: I think you should place it above the impl block

Comment (Erigara) [reply data_model/src/asset.rs:422]: My intention to capture relation between builder and target type. 
To be able to express smt like this:
```rust
impl Identifiable for NewRole {
    type Id =  <<Self as Builder>::Target as Identifiable>::Id;
}
```
If we make `Builder` trait available only under "mutable_api" feature we lose this ability.
I can do it both ways, what do you think?

Comment (mversic) [reply data_model/src/asset.rs:422]:
```rs
impl Identifiable for NewRole { 
    type Id = <<Self as Builder>::Target as Identifiable>::Id; 
}
```
is essentially the same thing as
```rs
impl Identifiable for NewRole { 
    type Id = <Role as Identifiable>::Id; 
}
```
after the compiler fills in the types

Comment (Erigara) [reply data_model/src/asset.rs:422]: Ok, i will make `Builder` only available under "mutable_api", thank you for the suggestion.

**Review by mversic — DISMISSED** [review]

**Review by mversic — DISMISSED** [review]

File: data_model/src/lib.rs [hunk 2]
```diff
@@ -745,6 +745,16 @@ pub trait Registered: Identifiable {
...
 
+/// Trait for objects used to create target objects
+#[cfg(feature = "mutable_api")]
+pub trait Builder {
+    /// Constructed type
+    type Target;
+
+    /// Construct [`Self::Target`]
+    fn build(self) -> Self::Target;
+}
```
Comment (s8sato) [inline data_model/src/lib.rs:756]: What do you think about this:

```suggestion
pub trait Registrable {
    /// Constructed type
    type Target: Identifiable;

    /// Identification of [`Self::Target`]
    fn target_id(&self) -> <Self::Target as Identifiable>::Id;

    /// Construct [`Self::Target`]
    fn build(self) -> Self::Target;
}
```

File: data_model/src/domain.rs [hunk 3]
```diff
@@ -77,13 +77,30 @@ impl From<GenesisDomain> for Domain {
 #[display(fmt = "[{id}]")]
 pub struct NewDomain {
     /// The identification associated to the domain builder.
-    id: <Domain as Identifiable>::Id,
+    id: <Self as Identifiable>::Id,
```
Comment (s8sato) [inline data_model/src/domain.rs:80]: and this:

```suggestion
    id: <<Self as Registrable>::Target as Identifiable>::Id,
```

**Review by s8sato — DISMISSED** [review]

File: data_model/src/lib.rs (hunk 2 above)
Comment (mversic) [reply data_model/src/lib.rs:756]: it does make sense to combine them into one trait. `Builder` and `Register` always go together, at least at the moment. I don't think we need `fn target_id` though

Comment (Erigara) [reply data_model/src/lib.rs:756]: As i find out there some cases where type implement `Registered<With = Self>` (Trigger, Peer, Account).
In this case having single trait `Builder + Registered` seems inconvenient.
However @s8sato suggestion is still applicable.

File: client/tests/integration/queries/role.rs [hunk 4]
```diff
@@ -91,7 +91,7 @@ fn find_role_by_id() -> Result<()> {
 
     let found_role = test_client.request(client::role::by_id(role_id))?;
 
-    assert_eq!(found_role.id(), new_role.id());
+    assert_eq!(found_role.id(), new_role.target_id());
```
Comment (mversic) [inline client/tests/integration/queries/role.rs:94]:
```suggestion
    assert_eq!(found_role.id(), new_role.build().id());
```

File: data_model/src/lib.rs [hunk 5]
```diff
@@ -741,6 +741,19 @@ pub trait Registered: Identifiable {
     type With: Into<RegistrableBox>;
 }
 
+/// Trait for proxy objects used for registration.
+pub trait Registrable {
+    /// Constructed type
+    type Target: Identifiable + Registered<With = Self>;
+
+    /// Identification of [`Self::Target`]
+    fn target_id(&self) -> &<Self::Target as Identifiable>::Id;
```
Comment (mversic) [inline data_model/src/lib.rs:750]: this method is redundant. You will never read the `id` of a builder, i.e. `NewXXX` structure. You always know which id you stored into the builder. I gave you a [suggestion](https://github.com/hyperledger/iroha/pull/2503/files#r926764681) in one place that it's used

Comment (mversic) [inline data_model/src/lib.rs:750]: this method is redundant. You will never read the id of a builder, i.e. NewXXX structure. You always know which id you stored into the builder. Here is a [suggestion](https://github.com/hyperledger/iroha/pull/2503/files#r926764681) in one place that it's used outside this crate

File: data_model/src/lib.rs [hunk 6]
```diff
@@ -261,10 +261,10 @@ pub enum IdentifiableBox {
...
     fn id_box(&self) -> IdBox {
         match self {
-            IdentifiableBox::NewDomain(a) => a.id().clone().into(),
-            IdentifiableBox::NewAccount(a) => a.id().clone().into(),
-            IdentifiableBox::NewAssetDefinition(a) => a.id().clone().into(),
-            IdentifiableBox::NewRole(a) => a.id().clone().into(),
+            IdentifiableBox::NewDomain(a) => a.target_id().clone().into(),
+            IdentifiableBox::NewAccount(a) => a.target_id().clone().into(),
+            IdentifiableBox::NewAssetDefinition(a) => a.target_id().clone().into(),
+            IdentifiableBox::NewRole(a) => a.target_id().clone().into(),
```
Comment (mversic) [inline data_model/src/lib.rs:267]: you can access the field directly with if you make it `pub(crate)` which is ok

File: data_model/src/lib.rs [hunk 7]
```diff
@@ -261,10 +261,10 @@ pub enum IdentifiableBox {
 impl IdentifiableBox {
     fn id_box(&self) -> IdBox {
         match self {
-            IdentifiableBox::NewDomain(a) => a.id().clone().into(),
-            IdentifiableBox::NewAccount(a) => a.id().clone().into(),
-            IdentifiableBox::NewAssetDefinition(a) => a.id().clone().into(),
-            IdentifiableBox::NewRole(a) => a.id().clone().into(),
+            IdentifiableBox::NewDomain(a) => a.id.clone().into(),
```
Comment (mversic) [inline data_model/src/lib.rs:264]: this was my initial suggestion, I would have preferred inherent method, eh. :)

**Review by mversic — DISMISSED** [review]

File: data_model/src/role.rs [hunk 8]
```diff
@@ -126,15 +126,17 @@ impl Registered for Role {
     IntoSchema,
 )]
 pub struct NewRole {
-    inner: Role,
+    pub(crate) inner: Role,
 }
 
-impl Identifiable for NewRole {
-    type Id = <Role as Identifiable>::Id;
+impl Registrable for NewRole {
```
Comment (mversic) [inline data_model/src/role.rs:132]: I think entire trait can be `mutable_api`

Comment (mversic) [reply data_model/src/role.rs:132]: it just increases API surface for the client without any benefit

**Review by mversic — APPROVED** [review]

File: data_model/src/lib.rs [hunk 9]
```diff
@@ -741,6 +741,16 @@ pub trait Registered: Identifiable {
     type With: Into<RegistrableBox>;
 }
 
+/// Trait for proxy objects used for registration.
+#[cfg(feature = "mutable_api")]
+pub trait Registrable {
+    /// Constructed type
+    type Target: Identifiable + Registered<With = Self>;
```
Comment (mversic) [inline data_model/src/lib.rs:748]:
```suggestion
    type Target: Registered<With = Self>;
```

Seems that it's implied by `Registered`

Comment (mversic) [reply data_model/src/lib.rs:748]: or should we remove the bound from `Registered`?

Comment (Erigara) [reply data_model/src/lib.rs:748]: I think `Registered` is unnecessary, so we can remove it.

Comment (Erigara) [reply data_model/src/lib.rs:748]: Actually we can omit any trait bounds on `Target`, since they are not used.

Comment (mversic) [reply data_model/src/lib.rs]: Are you getting rid of `Registered`?

Comment (Erigara) [reply data_model/src/lib.rs]: Yes, in `Target` trait bounds.

**Review by s8sato — APPROVED** [review]

**Pull request merged**